rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -255,170 +255,178 @@ private boolean isValidationRunCompleted(ValidationResult validationResult) {
+                return result;
             }
 
-            temporary.setLocation(new ValidationLocation(manifestUri));
-            manifest.validate(manifestUri.toASCIIString(), context, x509Crl, manifest.getCrlUri(), validationConfig.validationOptions(), temporary);
-            if (temporary.hasFailureForCurrentLocation()) {
-                return;
+            result.setLocation(new ValidationLocation(manifestUri));
+            manifest.validate(manifestUri.toASCIIString(), context, x509Crl, manifest.getCrlUri(), validationConfig.validationOptions(), result);
+            if (result.hasFailureForCurrentLocation()) {
+                return result;
             }
-            accumulator.add(manifestObject.get().key(), manifest, ImmutableSortedSet.of(temporary.getCurrentLocation().getName()));
-
-            manifest.getFiles().entrySet().stream()
-                .flatMap(entry -> getManifestEntry(manifestUri, entry))
-                .parallel()
-                .flatMap(tuple -> getCertificateRepositoryObjectValidationContext(trustAnchor, context, accumulator, crlUri, x509Crl, tuple))
-                .map(tuple -> {
-                    final ValidationResult vr = tuple.v2();
-                    validateCertificateAuthority(trustAnchor, registeredRepositories, tuple.v1(), vr, accumulator);
-                    return vr;
-                })
-                .forEachOrdered(temporary::addAll);
-        } catch (ManifestEntryException e){
-            temporary.addAll(e.getVr());
-        }
-        catch (Exception e) {
-            validationResult.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
+
+            Tuple2<ValidationResult, List<CertificateRepositoryObjectValidationContext>> validatedManifestEntries
+                    = validateManifestEntries(trustAnchor, context, manifestUri, manifestObject.get(), manifest, crlUri, x509Crl, accumulator);
+            result.addAll(validatedManifestEntries.v1());
+
+            validatedManifestEntries.v2().stream().map(childContext ->
+                    validateCertificateAuthority(trustAnchor, registeredRepositories, childContext, accumulator)
+            ).forEachOrdered(result::addAll);
+        } catch (Exception e) {
+            result.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
         }
+        return result;
+    }
 
-        finally {
-            validationResult.addAll(temporary);
+    private Tuple2<ValidationResult, List<CertificateRepositoryObjectValidationContext>> validateManifestEntries(
+            TrustAnchor trustAnchor,
+            CertificateRepositoryObjectValidationContext context,
+            URI manifestUri,
+            RpkiObject manifestObject,
+            ManifestCms manifest,
+            URI crlUri,
+            X509Crl x509Crl,
+            Accumulator accumulator
+    ) {
+        ValidationResult temporary = newValidationResult(context.getLocation());
```

Comment:
Personal preference, variable rename into `result` in the `validateCertificateAuthority` above makes it clear that it will be the result rather than `temporary`